src/mpi/session/mpir_pset.c

```diff
@@ -0,0 +1,201 @@
+ * @param   pset_name   Name of the pset
+ * @param   pset        Returns the found pset
+ * @return  int         MPI_SUCCESS or MPI_ERR_OTHER (not found)
+ */
+static
+int pset_find_by_name(UT_array * parray, const char *pset_name, MPIR_Pset ** pset)
+{
+    int ret = MPI_ERR_OTHER;
+    MPIR_Pset *p = NULL;
+    *pset = NULL;
+    for (unsigned i = 0; i < utarray_len(parray); i++) {
+        p = (MPIR_Pset *) utarray_eltptr(parray, i);
+        if (strncasecmp(pset_name, p->uri, MAX(strlen(pset_name), strlen(p->uri))) == 0) {
+            ret = MPI_SUCCESS;
+            *pset = p;
+            break;
+        }
+    }
+    return ret;
+}
+
+int MPIR_Pset_by_name(MPIR_Pset_array * pset_array, const char *pset_name, MPIR_Pset ** pset)
+{
+    int ret;
+
+    PSET_CS_ENTER(pset_array->mutex);
+
+    if (pset_name != NULL) {
+        ret = pset_find_by_name(pset_array->parray, pset_name, pset);
+    } else {
+        ret = MPI_ERR_OTHER;
+    }
+
+    PSET_CS_EXIT(pset_array->mutex);
+
+    return ret;
+}
+
+int MPIR_Pset_by_idx(MPIR_Pset_array * pset_array, int idx, MPIR_Pset ** pset)
+{
+    int ret = MPI_SUCCESS;
+
+    PSET_CS_ENTER(pset_array->mutex);
+
+    unsigned len = utarray_len(pset_array->parray);
+
+    /* Overflow check between unsigned and int data types */
+    MPIR_Assert(len <= INT_MAX);
+
+    if (idx >= 0 && idx < (int) len) {
+        *pset = (MPIR_Pset *) utarray_eltptr(pset_array->parray, idx);
+    } else {
+        ret = MPI_ERR_OTHER;
+    }
+
+    PSET_CS_EXIT(pset_array->mutex);
+
+    return ret;
+}
+
+void MPIR_Pset_array_init(MPIR_Pset_array ** pset_array, MPID_Thread_mutex_t * mutex)
+{
+    (*pset_array) = MPL_malloc(sizeof(MPIR_Pset_array), MPL_MEM_SESSION);
+
+    utarray_new((*pset_array)->parray, &pset_array_icd, MPL_MEM_SESSION);
+
+    (*pset_array)->mutex = mutex;
+    if ((*pset_array)->mutex) {
+        /* Init the mutex to protect the pset array from concurrent accesses */
+        int thr_err;
+        MPID_Thread_mutex_create((*pset_array)->mutex, &thr_err);
+        MPIR_Assert(thr_err == MPI_SUCCESS);
+    }
+}
+
+void MPIR_Pset_array_destroy(MPIR_Pset_array * pset_array)
+{
+    utarray_free(pset_array->parray);
+    if (pset_array->mutex) {
+        int thr_err;
+        MPID_Thread_mutex_destroy(pset_array->mutex, &thr_err);
+        MPIR_Assert(thr_err == MPI_SUCCESS);
+    }
+    MPL_free(pset_array);
+}
+
+int MPIR_Pset_array_add(MPIR_Pset_array * pset_array, MPIR_Pset * pset)
+{
+    int ret;
+    MPIR_Pset *p;
+
+    PSET_CS_ENTER(pset_array->mutex);
+
+    if (pset_find_by_name(pset_array->parray, pset->uri, &p) == MPI_ERR_OTHER) {
+        /* Pset with uri NOT found in the parray */
+        utarray_push_back(pset_array->parray, pset, MPL_MEM_SESSION);
+        ret = MPI_SUCCESS;
+    } else {
+        ret = MPI_ERR_OTHER;
+    }
+
+    PSET_CS_EXIT(pset_array->mutex);
+
+    return ret;
+}
+
+int MPIR_Pset_array_invalidate(MPIR_Pset_array * pset_array, char *pset_name)
```

Comment:
Since different session may have different psets and have different ideas of what is valid and invalid, I think each session should maintain an array of global pset pointers that are always valid within the session.

Reply:
all the sessions will have consistent picture due to malleability won't happen unless there is a gap where no session exist